### ISSUE TITLE
It should be possible to add a comment to both an inserted and deleted line of a diff

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-server#12875]
This is a follow-on from https://github.com/mozilla/addons-code-manager/issues/989#issuecomment-553988392.

Currently the Comment API only supports adding comments to lines in the "new" version of a diff. This means that, when displaying a unified diff, as we do in code-manager, a user would only be able to comment on the inserted (or green) lines, and not on the deleted (or red) lines. Because any change is represented by both an insert and a delete, it seems that a user should be able to comment on either (or even both) lines.

@wagnerand Could you comment on whether you agree that this should be possible?

**Comment by EnTeQuAk** [comment]
Alrighty, so we had a lengthy discussion on IRC about this.

Just to rephrase our status quo and why this is fairly complex (not to fix but to find a working workflow):

We are currently doing reviews (leaving comments) in both views, the browse view and the diff view. This makes it difficult to explicitly require a second version for context for a specific comment.

The question here is: Do we expect a comment that you filed when reviewing version 8 against version 7 to appear when browsing through version 8?

The answer is probably yes, especially since it can very well happen that a new version can be reviewed entirely in browse-mode (e.g when the version is a full rewrite or the amount of changes is big enough to justify a full review).

This still kinda follows the mantra: We are always commenting in the context of the most recent version. So, maybe… this means we don't really need to comment on a deleted line? And if we do, we'll comment on the whole file instead and add the context on our own?

**Comment by bobsilverberg** [comment]
IMO, it would break the workflow to disallow commenting on a deleted line, because we are displaying a unified diff. Imagine doing code reviews in GitHub and having that limitation in place. You could never comment directly on the "old" side of any change. If all are in agreement that we can introduce this limitation and only allow commenting on inserted lines in a diff, then I agree that would solve the issue, but I am wary of introducing such a limitation.

I do understand that the desire to see all comments when browsing version 8 (as in @EnTeQuAk's example above) introduces some problems for which we do not have a  solution (e.g., where would a comment on a deleted line appear? How would we maintain context?). I also think, however, that disallowing comments on deleted lines might be too much of a limitation to be acceptable.

I suppose a compromise would be to only show comments on *inserted* lines when in browse mode, but I can also see how that inconsistency could be confusing to users.

@kumar303 / @wagnerand What do you think?

**Comment by wagnerand** [comment]
I agree with @bobsilverberg. The limitation could become an issue at some point very quickly. On the other hand, we should make sure that any and every comment that is made will be shown both in browse mode and in diff mode.

**Comment by EnTeQuAk** [comment]
I do agree with @bobsilverberg here too.

So, while in compare mode maintaining context and avoiding confusion to the user is already "given" because all the information about the deleted line is there.

That means we only need a solution for the browse mode. Now, while in browse mode we technically don't have any information about any "deleted lines" because we don't compare a version to another one. What I'd propose is, all comments are saved for the "currently reviewing" version anyway, that means that we can store comments on deleted lines there too but if we do have the context store the version compared against too.

Once we have that we then need to store the information of whether that comment is for a deleted line or not. Once #12790 has been done that can be done automatically - until then I think it should be okay for the frontend to push that information to the server instead?

If we then have both, two versions and the "state" (deleted, etc) for a line of a specific comment saved we can safely show them on the currently reviewing version. How we're displaying them is another issue, maybe just put them on the line number where something has been deleted and make very very clear in the comment-box that this comment relates to a line when compared to the previous (or whichever) version?

With all that the only compromise would be, that when in browse-mode for the previous version (with the deleted line) we won't be able to show a comment there too (as the comments are saved on the currently reviewing version). (while technically possible… I personally think that'd be too confusing or am I overreacting here?).

Summary:

* Add the ability to store a "context"-version (the version we compare against) to the comments API
* Add the (temporary) ability to store a "state" (e.g deleted-line) for a comment (although that information will only be useful if we show it to the user, if we don't, we won't need it)

Once that has been done the frontend will get all the necessary information to show that additional information when in browse mode.

Does that make sense?

**Comment by wagnerand** [comment]
> with all that the only compromise would be, that when in browse-mode for the previous version (with the deleted line) we won't be able to show a comment there too

Since those annotations are made in context to a specific line, the comment should be shown in this case.

**Comment by EnTeQuAk** [comment]
> Since those annotations are made in context to a specific line, the comment should be shown in this case.

And while thinking about that in the meantime, it's technically possible anyway so it certainly just makes sense to show it.

Everything else from my comment above applies though.

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDSRV-23
